Add a scripted similarity.

jpountz opened the pull request:
The goal of this similarity is to help users who would like to keep the
functionality of the `tf-idf` similarity that we want to remove, or to allow
for specific use-cases (disabling idf, disabling tf, disabling length norm,
etc.) to not have to build a custom plugin and familiarize with the low-level
Lucene API.

This is a work-in-progress that needs more tests, but I would like to get
early feedback about the impact of this PR on the scripting API and whether
I should do things differently.

jdconrad commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java:
```diff
@@ -660,7 +660,7 @@ private void addElements() {
 
     private void addStruct(final String name, final Class<?> clazz) {
-        if (!name.matches("^[_a-zA-Z][\\.,_a-zA-Z0-9]*$")) {
+        if (!name.matches("^[_a-zA-Z][\\.,_a-zA-Z0-9\\$]*$")) {
```
We actually don't allow types in Painless to have the '$' as part of the type.  If I remember correctly the history here is that we have internal variables that have '$' that we don't want to have conflicts.  See later comment for a recommendation on how to resolve this.

If we want to change this we'd also have to modify the existing lexer/parser which I'm not sure is worth it with the workaround I mention in a later comment.

jdconrad commented on modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.txt:
```diff
@@ -165,3 +165,30 @@ class org.elasticsearch.search.lookup.FieldLookup -> org.elasticsearch.search.lo
   boolean isEmpty()
 }
+
+class org.elasticsearch.index.similarity.ScriptedSimilarity$Query -> org.elasticsearch.index.similarity.ScriptedSimilarity$Query extends Object {
```
For the whitelisted classes you can change the type name to be like the following:

```class org.elasticsearch.index.similarity.ScriptedSimilarity.Query -> org.elasticsearch.index.similarity.ScriptedSimilarity$Query extends Object {```

Note the '$' got changed to '.' in the Painless name of the type (first piece) while it still represents the appropriate Java class (second piece).  All the rest of the defined types will work this way too.

rjernst commented on core/src/main/java/org/elasticsearch/index/similarity/IBSimilarityProvider.java:
```diff
@@ -65,7 +66,7 @@
     private final IBSimilarity similarity;
 
-    public IBSimilarityProvider(String name, Settings settings, Settings indexSettings) {
+    public IBSimilarityProvider(String name, Settings settings, Settings indexSettings, ScriptService scriptService) {
```
Can we avoid adding ScriptService to the ctor of every provider? It should only be needed for the new one?

rjernst commented on core/src/main/java/org/elasticsearch/index/similarity/ScriptedSimilarity.java:
```diff
@@ -0,0 +1,282 @@
+        }
+    }
+
+    public static class Stats {
```
Some javadocs on these please?

rjernst commented on core/src/main/java/org/elasticsearch/index/similarity/ScriptedSimilarity.java:
```diff
@@ -0,0 +1,282 @@
+public final class ScriptedSimilarity extends Similarity {
+
+    private final String scriptString;
+    private final Supplier<ExecutableScript> scriptSupplier;
```
We should be using a new script context here. Then this can be SimilarityScript.Factory.  The new context can return `float` directly, and take Stats as an arg.

rjernst commented on core/src/main/java/org/elasticsearch/index/similarity/ScriptedSimilarityProvider.java:
```diff
@@ -0,0 +1,65 @@
+        super(name);
+        boolean discountOverlaps = settings.getAsBoolean("discount_overlaps", true);
+        String lang = settings.get("lang", Script.DEFAULT_SCRIPT_LANG);
+        String source = settings.get("source");
```
You should be able to use Script.parse? Or we should make it so you can. We should not have to duplicate all this (it is something that has been a pain point in ingest scripts as it must be kept synchronized with other script parsing code).

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/GenericElasticsearchScript.java:
```diff
@@ -30,10 +31,12 @@
 
     public abstract boolean needs_score();
     public abstract boolean needsCtx();
+    public abstract boolean needsStats();
```
You shoudl not need this if you create a new context, SimilarityScript.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/ScriptImpl.java:
```diff
@@ -93,6 +96,9 @@
     @Override
     public void setNextVar(final String name, final Object value) {
         variables.put(name, value);
+        if (script.needsStats() && "stats".equals(name)) {
```
Should not be needed

jpountz commented:
Thanks for the notes about how to use the new context API, I know I was doing something wrong but I wasn't sure what I was supposed to do instead. I addressed all comments, would you mind having another look? I'm especially interested to know whether there are things that could be done more efficiently as I'd really like to be as efficient as a similarity plugin.

rjernst commented on core/src/main/java/org/elasticsearch/script/SimilarityScript.java:
```diff
@@ -0,0 +1,39 @@
+public abstract class SimilarityScript  {
+
+    /** Compute the score. */
+    public abstract double execute(double weight, ScriptedSimilarity.Query query,
```
Why not have this return float since you just cast it to that anyways when it is used? Then the script writer can make the explicit choice as to how the precision is reduced to float?

jpountz replied:
I think scripts would use doubles as an intermediate representation almost all the time anyway given that functions like log or sqrt produce doubles, so I felt like returning a double and casting on the elasticsearch side would save one cast from every similarity script.

Also we might switch to doubles for scores in the future (https://issues.apache.org/jira/browse/LUCENE-7517).

rjernst replied:
That makes sense.

rjernst commented on core/src/main/java/org/elasticsearch/index/similarity/ScriptedSimilarity.java:
```diff
@@ -0,0 +1,284 @@
+
+    final String initScriptString;
+    final String scriptString;
+    final Supplier<SimilarityScript> initScriptSupplier;
```
Can we just use SimilarityScript.Factory here?

rjernst commented on core/src/main/java/org/elasticsearch/index/similarity/ScriptedSimilarity.java:
```diff
@@ -0,0 +1,284 @@
+    /** Statistics that are specific to a given field. */
+    public static class Field {
+        final long docCount;
+        final long sumDocFreq;
```
These could be private since the getters are public? Or the getters could be removed and just make these public final?

rjernst commented on core/src/main/java/org/elasticsearch/index/similarity/ScriptedSimilarityProvider.java:
```diff
@@ -0,0 +1,56 @@
+        SimilarityScript.Factory initScriptFactory = null;
+        if (initScriptSettings.isEmpty() == false) {
+            initScript = Script.parse(initScriptSettings);
+            initScriptFactory = scriptService.compile(initScript, SimilarityScript.CONTEXT);
```
I think this should be a different script context. Then its execute can only take in query, field and term. I also think the name "init_script" is too generic? Maybe term_weight_script?

jpountz commented:
@rjernst Thanks for the review, I pushed a new commit.

jpountz commented:
@rjernst Could you take another look please?

rjernst commented on core/src/main/java/org/elasticsearch/index/similarity/ScriptedSimilarity.java:
```diff
@@ -0,0 +1,284 @@
+ */
+public final class ScriptedSimilarity extends Similarity {
+
+    final String weightScriptString;
```
nit: can we use the suffix "source" instead of "string" for these scripts? That matches what we call it now in scripting code.

rjernst commented on core/src/main/java/org/elasticsearch/script/SimilarityWeightScript.java:
```diff
@@ -0,0 +1,39 @@
+public abstract class SimilarityWeightScript  {
+
+    /** Compute the weight. */
+    public abstract double execute(ScriptedSimilarity.Query query, ScriptedSimilarity.Field field,
```
I'm not sure if the javadocs here are the right place, but we should document the parameters for users somewhere?
